graphviz/jupyter_integration.py

```diff
@@ -1,12 +1,70 @@
 """Display rendered graph as SVG in Jupyter Notebooks and QtConsole."""
-
 from . import piping
 
 __all__ = ['JupyterSvgIntegration']
 
+_default_jupyter_representation = 'image/svg+xml'
+
+_jupyter_repr_functions = {
+    'image/svg+xml': lambda graph: graph.pipe(format='svg',
+                                              encoding=graph._encoding),
+    'image/png': lambda graph: graph.pipe(format='png'),
+    'image/jpeg': lambda graph: graph.pipe(format='jpeg'),
+
+    # The following does not work yet. Not clear why, because the documentation
+    # has only few information about them.
+    # 'application/json': lambda graph: json.loads(
+    #     graph.pipe(format='json', encoding=graph._encoding)),
+    # 'application/pdf': lambda graph: graph.pipe(format='pdf'),
+
+    # No reasonable representation
+    # 'text/plain': None,
+    # 'text/latex': None,
+    # 'text/html': None,
+    # 'application/javascript': None,
+    # 'text/markdown': None,
+}
+
+JUPYTER_REPRESENTATIONS = set(_jupyter_repr_functions.keys())
```

Comment:
If I get it right, the methods will later be on the graph object, so they should be verbose, i.e. `_repr_image_svg_xml`.
When you see that, you will see the similarity with the suggestion from IPython `_repr_svg_` and it would then be more natural to use `_repr_svg_`. Sadly, the existence of `_repr_svg_` will disable `_repr_mimebundle_`.

So the private methods will differ from the suggested ones, and that feels slightly strange.

I just wanted to clarify this. With a "global" dict, it would also be slightly easier to add custom reprs (I could add the PDF to svg example to the docs, which I don't want to add to the source code, because it probably only work on linux).

Reply:
As these methods are properly private/internal (as oposed to `__repr_svg_` and `_repr_mimebundle_` which are pseudo-private), I think the naming is really only relevant for developers of this library (not for end users).

`_repr_image_svg_xml` SG.

>  With a "global" dict

Not sure if I understand what exactly you mean with this, can you expand a bit?

In case your goal is to allow end users to implement formats for which we don't provide a default implementation I guess we would need to change the logic for putting together `mimebundle` above to ` if format in reprs and hasattr(self, method_name)`. However, I am not convinced yet that we should support that.

Reply:
> Not sure if I understand what exactly you mean with this, can you expand a bit?

My motivation to write the issue was, that I had a broken graphviz installation and wanted to change, how jupyter displays the graphs. Since it was for teaching, I searched for a solution inside a Jupyter Notebook. e.g. put
```python
import graphviz
def display_fix(...): ...
graphviz.jupyter_integration.JUPYTER_REPRESENTATIONS['graphviz_vX.Y.Z_fix'] = display_fix()
graphviz.jupyter_integration.set_default_jupyter_representation('graphviz_vX.Y.Z_fix')
```
in the first cell. Now where you mention it
```python
import graphviz
def display_fix(...): ...
graphviz.jupyter_integration.JupyterSvgIntegration._repr_mimebundle_ = display_fix
```
will also do the job. More hacky, but does job.

I will follow you suggestion.

Reply:
I see, thanks.

If one would need to monkeypatch, I think one could also do it like this:

```python
class JupyterIntegrationPatch:
    def _repr_mimebundle_(sef, ...):
          ...

 class Graph(JupyterIntegrationPatch, graphviz.Graph):
    ...

 class Digraph(JupyterIntegrationPatch, graphviz.Digraph):
    ...

 class Source(JupyterIntegrationPatch, graphviz.Source):
    ...
```

Reply:
That depends, how much control you have. When you use code, that uses `graphviz`, you cannot replace the classes.
In my case, I just want to have a notebook cell, where I modify, what `graphviz` uses for display in a single cell and don't want to touch the external code.

But with the current implementation, I am able to use a workaround for the bug. So its fine for me.

Reply:
Agreed.

Just to make sure we are on the same page here: when you do something like below, you are monkey-patching private internals of `graphviz`, so if this breaks (e.g. because we rearrange the code or rename internal classes), it's completely your own responsibility (i.e. we won't consider this a bug in `graphviz`).

`graphviz.jupyter_integration.JupyterSvgIntegration._repr_mimebundle_ = display_fix`

Reply:
Yes, such a code is hacking the internals and may break with a new release. But it is a fix for graphviz (i.e. not Python graphviz) installed on the Linux machine, that I cannot update. So such a hack is in my mind ok and does not need public an API.

Reply:
Kewl, thanks.